File: packages/constants/src/app-data.ts
/**
 * Where Tironian stores things on a machine.
 *
 * Tironian owns exactly one application-data root. Every trusted app it runs or
 * admits receives one directory below it and owns everything inside, partitioned
 * by an identifier the external authority owns and never reuses. See ADR-0201.
 *
 * Three parties choose names along that path, and each function below is one
 * hand-off between two of them: Tironian names the root and its own
 * directories, an app names everything in its directory, and an external
 * authority names a partition. `apps/` and the partition-kind directory exist
 * because a directory whose next name is chosen by somebody else cannot be
 * defended by the party that would have to defend it. There is no level here
 * that is not one of those hand-offs.
 *
 * These are pure functions over strings and a grammar. There is no store, no
 * handle, no registry of app directories, and no lifecycle: allocating a place
 * is naming it, not owning a store, and the host never creates, opens, reads, or
 * reclaims anything below the directory it names.
 */

import { homedir } from 'node:os';
import { isAbsolute, join } from 'node:path';

/**
 * The `app.tironian` bundle identity, which is what names the root on every
 * platform. It has to equal `identifier` in
 * `apps/desktop/src-tauri/tauri.conf.json`, because the desktop host resolves
 * the same directory through Tauri; `app-data.test.ts` pins the two together.
 */
export const TIRONIAN_BUNDLE_IDENTIFIER = 'app.tironian';

/**
 * The one grammar for an app id, shared with catalog admission.
 *
 * An app id names a place, and two issuers name into that one space: admission
 * issues one when it accepts a folder. For an admitted app, that id is the
 * reverse-domain workspace id the folder declares (ADR-0210); the composition
 * root also issues ids for the engines it composes. The grammar has one
 * definition because those ids share one identifier space; a second copy of this
 * pattern is how they would drift apart.
 *
 * Dots are admitted because an admitted app's id *is* its reverse-domain
 * workspace id, and bare labels stay legal so the composed ids (`local-mail`,
 * `local-books`) keep the directories they already own.
 *
 * The first and last character must be alphanumeric, and that is load-bearing
 * rather than tidy: {@link appDataDir} joins an id onto the one data root, so a
 * grammar admitting `.` or `..` would hand a caller a path out of the root, and
 * one admitting a leading dot would let an app hide as a dotfile.
 */
const APP_ID_PATTERN = /^[a-z0-9](?:[a-z0-9.-]*[a-z0-9])?$/;

export function isAppId(value: string): boolean {
	return APP_ID_PATTERN.test(value);
}

/**
 * App ids the composition root has already spent, which admission therefore
 * cannot re-issue to a folder.
 *
 * This is not the set of apps that own a directory: every trusted app owns one
 * (ADR-0201). It is the set of ids that arrive through no catalog, because their
 * apps are composed directly or ship as a standalone CLI, so admission has no
 * other way to know the names are taken. A folder named `local-mail` would
 * otherwise be admitted as a second claimant on the directory Local Mail's
 * credentials and intent store already sit in.
 *
 * An app id is deliberately not a surface id: Local Books has no launchable
 * surface at all, and coupling a mailbox's location to a name Home's launcher
 * owns would let a surface rename strand data (ADR-0201).
 */
export const COMPOSED_APP_IDS = ['local-mail', 'local-books'] as const;

export type ComposedAppId = (typeof COMPOSED_APP_IDS)[number];

/**
 * The ambient inputs the root is computed from, passed as a value so the
 * platform table below is a unit test rather than a machine you have to own.
 * Defaults to this process.
 */
export type DataRootSystem = {
	env: Record<string, string | undefined>;
	platform: string;
	homeDir: string;
};

/**
 * The one Tironian application-data root. `TIRONIAN_DATA_DIR` wins, for tests
 * and for a person who wants their data elsewhere; an empty value counts as
 * unset and a relative one is refused, for the same reason a relative
 * `XDG_DATA_HOME` is ignored below.
 *
 * Otherwise this reproduces what the desktop host resolves through Tauri, and
 * the equality is the point: a host and a CLI that disagree here write to two
 * different mailboxes. Tauri 2.11's `app_data_dir()` is `dirs::data_dir()`
 * joined with the bundle identifier, and `dirs` 6.0 resolves `data_dir()` as
 * `$HOME/Library/Application Support` on macOS, `$XDG_DATA_HOME` *only when it
 * is absolute* and `$HOME/.local/share` otherwise on Linux and other Unix, and
 * roaming `%APPDATA%` on Windows.
 *
 * Two of those clauses correct what `apps/local-mail` and `apps/local-books`
 * each do today: both honour a relative `XDG_DATA_HOME`, and neither has a
 * Windows branch, so a Windows install lands in `%USERPROFILE%\.local\share`.
 */
export function tironianDataRoot(
	system: DataRootSystem = {
		env: process.env,
		platform: process.platform,
		homeDir: homedir(),
	},
): string {
	const override = system.env.TIRONIAN_DATA_DIR;
	if (override && override.length > 0) {
		// A relative override would resolve against the working directory, so a CLI
		// run from two places would see two roots while the desktop host saw a
		// third: the exact drift a relative `XDG_DATA_HOME` is ignored for. It is
		// refused rather than resolved because that would read a fourth ambient
		// input this function deliberately does not take.
		if (!isAbsolute(override)) {
			throw new Error(
				`TIRONIAN_DATA_DIR must be an absolute path, not ${JSON.stringify(override)}.`,
			);
		}
		return override;
	}
	return join(dataDir(system), TIRONIAN_BUNDLE_IDENTIFIER);
}

/**
 * The one human-facing folder: where a person and an agent read this data.
 *
 * Deliberately not {@link tironianDataRoot}. That directory is machinery, is
 * explicitly not an inter-app API, and is where a continuous producer with no
 * consumer went to die once already (ADR-0010). This one is a place you `cd`
 * into, so it takes the shape every tool in its category converged on
 * independently: the home directory, capitalized, no dot (`~/Dropbox`,
 * `~/OneDrive`, and the same under `%USERPROFILE%` on Windows).
 *
 * Overridable, because a person may keep it elsewhere. The default has to stay
 * typeable, because "point your agent at `~/Tironian`" is the whole product
 * (ADR-0207).
 */
export function tironianFolderRoot(
	system: Pick<DataRootSystem, 'env' | 'homeDir'> = {
		env: process.env,
		homeDir: homedir(),
	},
): string {
	const override = system.env.TIRONIAN_FOLDER_DIR;
	if (override && override.length > 0) {
		// Refused rather than resolved, for the same reason the data root refuses
		// one: a relative path means the host and a CLI disagree about where the
		// folder is, which is the drift this function exists to prevent.
		if (!isAbsolute(override)) {
			throw new Error(
				`TIRONIAN_FOLDER_DIR must be an absolute path, not ${JSON.stringify(override)}.`,
			);
		}
		return override;
	}
	return join(system.homeDir, 'Tironian');
}

function dataDir({ env, platform, homeDir }: DataRootSystem): string {
	if (platform === 'darwin') {
		return join(homeDir, 'Library', 'Application Support');
	}
	if (platform === 'win32') {
		const appData = env.APPDATA;
		// `dirs` asks Windows for FOLDERID_RoamingAppData and yields nothing when
		// that fails, which Tauri turns into an error. Guessing a path here would
		// silently put a person's mail somewhere their host is not looking, so
		// this fails the same way rather than inventing a fallback.
		if (!appData || appData.length === 0) {
			throw new Error(
				'APPDATA is not set, so the Tironian data root cannot be resolved. Set TIRONIAN_DATA_DIR to name it explicitly.',
			);
		}
		return appData;
	}
	const xdg = env.XDG_DATA_HOME;
	// Absolute only, matching `dirs`. A relative XDG_DATA_HOME would otherwise
	// resolve against the working directory, so a CLI run from two places would
	// see two roots while the desktop host saw a third.
	if (xdg && isAbsolute(xdg)) return xdg;
	return join(homeDir, '.local', 'share');
}

/**
 * An app's one directory: `<root>/apps/<appId>`. The app owns everything below
 * the result and Tironian never looks inside it (ADR-0201, ADR-0193).
 *
 * `apps/` is where naming authority changes hands. Above it Tironian chooses
 * the names (`data`, `blobs`, `app-catalog`, and whatever it adds next); below
 * it an app does. One segment keeps a host directory added later from landing
 * on an app id, and it is the boundary the host's promise is stated against:
 * everything under `apps/` is somebody else's, all of it, by position.
 *
 * Allocation is nominal: this names a place and creates nothing. A directory
 * exists exactly when its owner writes into it, the same rule
 * {@link partitionDir} follows one level down, which is why every trusted app
 * having one costs nothing to run.
 *
 * The result is a string, injected at the owner's composition root the way the
 * sidecar already computes `join(root, 'data')` and `join(root, 'blobs')`. It is
 * deliberately not a capability: the bytes are not Tironian's to offer
 * (ADR-0181, ADR-0183).
 *
 * The id is validated rather than typed, because it comes from an open space:
 * an admitted app's id is a folder name (ADR-0179), not a literal this package
 * could enumerate. {@link COMPOSED_APP_IDS} is what keeps the two issuers from
 * naming the same directory, and it is checked at admission rather than here.
 */
export function appDataDir(root: string, appId: string): string {
	if (!isAppId(appId)) {
		throw new Error(
			`The app id ${JSON.stringify(appId)} cannot name a directory.`,
		);
	}
	return join(root, 'apps', appId);
}

/**
 * One partition of an app's directory: `<appDir>/<kind>/<partitionId>`.
 *
 * A partition holds everything scoped to one external account, company, or
 * tenancy, and `partitionId` must be an identifier that external authority
 * issues and never reuses.
 *
 * `kind` is the same hand-off as `apps/`, one altitude down. The app chooses
 * its root filenames (`credentials.json`, `provider.json`) and a provider
 * chooses partition ids, so one directory sits between the two namespaces
 * rather than a reserved-name rule the app would have to enforce against an
 * authority it does not control. The app picks the word (`accounts`,
 * `companies`), because only the app knows what it partitions.
 *
 * Both segments are validated as exactly one path component, which is the only
 * reason this exists rather than a bare `join`: a partition id arrives from a
 * provider callback or a command-line flag, and the Local Books call site this
 * replaced joined `realmId` verbatim.
 *
 * There is no acquisition protocol. A partition exists exactly when its
 * directory does; this function names one and creates nothing.
 */
export function partitionDir(
	appDir: string,
	kind: string,
	partitionId: string,
): string {
	assertOneSegment(kind, 'partition kind');
	assertOneSegment(partitionId, 'partition id');
	return join(appDir, kind, partitionId);
}

function assertOneSegment(segment: string, label: string): void {
	if (
		segment.length === 0 ||
		segment === '.' ||
		segment === '..' ||
		segment.includes('/') ||
		segment.includes('\\')
	) {
		throw new Error(
			`The ${label} ${JSON.stringify(segment)} cannot name a directory.`,
		);
	}
}

File: apps/desktop/src-tauri/src/app_data.rs
//! The one Tironian application-data root, resolved natively.
//!
//! `@tironian/constants/app-data` is the authority on this path, and the Bun
//! sidecar calls it directly (ADR-0201). Rust resolves it here for the one
//! consumer that cannot be handed the sidecar's answer: the staged-recording
//! blob store in [`crate::recorder::blob`], which writes `<root>/blobs` from
//! Rust. Who eventually tells the recorder where blobs live is an open
//! question; until it is settled this is the second implementation of one path,
//! and the equality is pinned by a test rather than by inspection.
//!
//! Mirroring the authority means mirroring its *inputs*, not just its result on
//! one config. This resolves the platform directory and joins the shared
//! identifier constant, exactly as the TypeScript does; it deliberately does not
//! use Tauri's `app_data_dir()`, whose answer moves with the bundle identifier
//! and so split dev builds away from the sidecar (see
//! [`TIRONIAN_BUNDLE_IDENTIFIER`]).
//!
//! The override is part of that equality and not an extra. Rust used to compute
//! the root and pass it to the sidecar in `TIRONIAN_DATA_DIR`, which meant an
//! ambient value was overwritten and could not split the two. The sidecar now
//! resolves its own root and honours the variable, so the recorder has to honour
//! it too: otherwise a person who moved their data would have recordings written
//! to one `blobs/` and served from another.

use std::ffi::OsStr;
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use tauri::{AppHandle, Manager, Runtime};

/// The one override for the one root, read by the sidecar, every CLI, and here.
const DATA_ROOT_OVERRIDE: &str = "TIRONIAN_DATA_DIR";

/// Mirrors `TIRONIAN_BUNDLE_IDENTIFIER` in `packages/constants/src/app-data.ts`,
/// which is the authority on this path.
///
/// Deliberately a constant rather than the running bundle's identifier. Tauri's
/// `app_data_dir()` is `data_dir()` joined with the *config's* identifier, and
/// `tauri.dev.conf.json` overrides that identifier to `app.tironian.dev` so the
/// dev build is a separate application (its own window state, webview data and
/// deep-link registration). The data root is not part of that separation: the
/// sidecar resolves `app.tironian` unconditionally, so a dev host built on
/// `app_data_dir()` wrote recordings to a `blobs/` the sidecar never served and
/// every dev dictation 404'd before it could be transcribed. Joining the shared
/// constant keeps the two implementations equal under every config, which is the
/// only property that matters here.
const TIRONIAN_BUNDLE_IDENTIFIER: &str = "app.tironian";

/// The root this machine's Tironian stores everything under.
pub fn tironian_data_root<R: Runtime>(app: &AppHandle<R>) -> Result<PathBuf> {
    resolve_data_root(std::env::var_os(DATA_ROOT_OVERRIDE).as_deref(), || {
        platform_root(app)
    })
}

/// The platform data directory joined with the shared bundle identifier: the
/// same two steps, in the same order, as `tironianDataRoot()` in TypeScript.
fn platform_root<R: Runtime>(app: &AppHandle<R>) -> Result<PathBuf> {
    Ok(app
        .path()
        .data_dir()
        .context("resolve the platform application-data directory")?
        .join(TIRONIAN_BUNDLE_IDENTIFIER))
}

/// Apply the override rule to a platform root.
///
/// An empty value counts as unset and a relative one is refused rather than
/// resolved, matching `tironianDataRoot` exactly: resolving a relative path
/// against the working directory would give a CLI run from two places two roots
/// while the desktop saw a third. The platform root is a closure so a set
/// override never needs it, which is what keeps a Windows machine without
/// `APPDATA` startable by naming the root explicitly.
fn resolve_data_root(
    override_value: Option<&OsStr>,
    platform_root: impl FnOnce() -> Result<PathBuf>,
) -> Result<PathBuf> {
    match override_value {
        Some(value) if !value.is_empty() => {
            let path = Path::new(value);
            if !path.is_absolute() {
                bail!("{DATA_ROOT_OVERRIDE} must be an absolute path, not {value:?}.");
            }
            Ok(path.to_path_buf())
        }
        _ => platform_root(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::process::Command;

    fn unreachable_root() -> Result<PathBuf> {
        panic!("the platform root must not be resolved when the override wins");
    }

    /// An absolute path spelled the way the running platform spells one.
    /// The rule under test is `Path::is_absolute`, and that is per-platform:
    /// `/tmp/tironian-test` is a relative path on Windows, so a Unix-only
    /// literal would test the refusal branch there instead of the override.
    const ABSOLUTE_OVERRIDE: &str = if cfg!(windows) {
        r"C:\tironian-test"
    } else {
        "/tmp/tironian-test"
    };

    #[test]
    fn an_absolute_override_wins_without_touching_the_platform_root() {
        assert_eq!(
            resolve_data_root(Some(OsStr::new(ABSOLUTE_OVERRIDE)), unreachable_root).unwrap(),
            PathBuf::from(ABSOLUTE_OVERRIDE)
        );
    }

    #[test]
    fn an_empty_override_counts_as_unset() {
        assert_eq!(
            resolve_data_root(Some(OsStr::new("")), || Ok(PathBuf::from("/platform"))).unwrap(),
            PathBuf::from("/platform")
        );
        assert_eq!(
            resolve_data_root(None, || Ok(PathBuf::from("/platform"))).unwrap(),
            PathBuf::from("/platform")
        );
    }

    #[test]
    fn a_relative_override_is_refused_rather_than_resolved() {
        let error = resolve_data_root(Some(OsStr::new("tmp/data")), unreachable_root).unwrap_err();
        assert!(error.to_string().contains("absolute"));
    }

    /// The identifier the running desktop is bundled under, read from the file
    /// Tauri itself reads, so this test cannot drift from the build.
    fn bundle_identifier() -> String {
        let config: serde_json::Value =
            serde_json::from_str(include_str!("../tauri.conf.json")).unwrap();
        config["identifier"].as_str().unwrap().to_string()
    }

    /// Ask Bun for `tironianDataRoot()` on this machine, with the override
    /// removed so the answer is the platform rule rather than an ambient value.
    fn typescript_data_root() -> PathBuf {
        let output = Command::new("bun")
            .arg("-e")
            .arg(concat!(
                "import { tironianDataRoot } from '@tironian/constants/app-data';",
                "process.stdout.write(tironianDataRoot());"
            ))
            .current_dir(concat!(env!("CARGO_MANIFEST_DIR"), "/.."))
            .env_remove(DATA_ROOT_OVERRIDE)
            .output()
            .expect(
                "run bun; the Tironian host is a Bun program and its tests assume the toolchain",
            );
        assert!(
            output.status.success(),
            "bun failed to resolve the data root: {}",
            String::from_utf8_lossy(&output.stderr)
        );
        PathBuf::from(String::from_utf8(output.stdout).unwrap())
    }

    /// The one equality neither side can check by reading the other.
    ///
    /// `@tironian/constants/app-data` transcribes what Tauri 2.11 and `dirs`
    /// 6.0 do; this runs both implementations on the machine the test is on and
    /// compares the answers. A `dirs` bump, a Tauri change, or an edited
    /// identifier fails here instead of silently splitting a person's data
    /// between a desktop host and a CLI (ADR-0201).
    #[test]
    fn the_native_root_equals_the_typescript_resolver() {
        let mut context = tauri::test::mock_context(tauri::test::noop_assets());
        context.config_mut().identifier = bundle_identifier();
        let app = tauri::test::mock_builder()
            .build(context)
            .expect("build a mock Tauri app for its path resolver");

        let native = resolve_data_root(None, || platform_root(app.handle())).unwrap();

        assert_eq!(native, typescript_data_root());
    }

    /// The dev bundle is a separate *application*, not a separate data root.
    ///
    /// `tauri.dev.conf.json` overrides the identifier to `app.tironian.dev`. When
    /// this resolver was built on `app_data_dir()` that override moved the root
    /// with it, so `bun dev:desktop` wrote recordings to
    /// `<data>/app.tironian.dev/blobs` while the sidecar served
    /// `<data>/app.tironian/blobs`, and every dev dictation failed with a 404
    /// before it could reach transcription. The equality test above could not
    /// catch it: it reads `tauri.conf.json`, so it only ever saw the production
    /// identifier. This pins the property that actually matters, that no config's
    /// identifier moves the root at all.
    #[test]
    fn the_dev_identifier_does_not_move_the_root() {
        let dev_config: serde_json::Value =
            serde_json::from_str(include_str!("../tauri.dev.conf.json")).unwrap();
        let dev_identifier = dev_config["identifier"].as_str().unwrap().to_string();
        assert_ne!(
            dev_identifier,
            bundle_identifier(),
            "this test is only meaningful while the dev config overrides the identifier"
        );

        let mut context = tauri::test::mock_context(tauri::test::noop_assets());
        context.config_mut().identifier = dev_identifier;
        let app = tauri::test::mock_builder()
            .build(context)
            .expect("build a mock Tauri app for its path resolver");

        assert_eq!(
            platform_root(app.handle()).unwrap(),
            typescript_data_root(),
            "the dev identifier must not split the recorder from the sidecar"
        );
    }
}
